data-science-onramp/data-cleaning/clean.py

```diff
@@ -0,0 +1,179 @@
+
+def station_name(name):
+    '''Replaces '/' with '&'.'''
+    if name:
+        return name.replace('/', '&')
+
+def user_type(user):
+    '''Converts user type to 'Subscriber' or 'Customer'.'''
+    if user:
+        if user.lower().startswith('sub'):
+            return 'Subscriber'
+        elif user.lower().startswith('cust'):
+            return 'Customer'
+        else:
+            return None
+
+def gender(s):
+    '''Converts gender to 'Male' or 'Female' or '''
+    if s:
+        if s.lower().startswith('m'):
+            return 'Male'
+        elif s.lower().startswith('f'):
+            return 'Female'
+        else:
+            return None
+
+def angle(a):
+    '''Converts DMS notation to angles.'''
+    if a:
+        dms = re.match('(-?\d*).(-?\d*)\'(-?\d*)"', a)
+        if dms:
+            return int(dms[1]) + int(dms[2])/60 + int(dms[3])/(60 * 60)
+        else:
+            try:
+                return float(a)
+            except ValueError:
+                return None
+
+def compute_time(duration, start, end):
+    '''Calculates duration, start time, and end time from each other if one value is null.'''
+    time_format = '%Y-%m-%dT%H:%M:%S'
+
+    # Convert to datetime/timedelta
+    if start:
+        if '.' in start:
+            start = start[:start.index('.')]
+        start = datetime.datetime.strptime(start, time_format)
+    if end:
+        if '.' in end:
+            end = end[:end.index('.')]
+        end = datetime.datetime.strptime(end, time_format)
+    if duration:
+        duration = datetime.timedelta(seconds=duration)
+
+    # Compute null value
+    if start and end and not duration:
+        duration = end - start
+    elif duration and end and not start:
+        start = end - duration
+    elif duration and start and not end:
+        end = start + duration
+
+    # Convert from datetime/timedelta
+    if duration:
+        duration = int(duration.total_seconds())
+    if start:
+        start = start.strftime(time_format)
+    if end:
+        end = end.strftime(time_format)
+
+    return (duration, start, end)
+        
+def compute_duration(duration, start, end):
+    '''Calculates duration from start and end time if null.'''
+    return compute_time(duration, start, end)[0]
+
+def compute_start(duration, start, end):
+    '''Calculates start time from duration and end time if null.'''
+    return compute_time(duration, start, end)[1]
+
+def compute_end(duration, start, end):
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
+
+def backfill(df):
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
+
+def main():
+    # Create a SparkSession under the name 'clean'. Viewable via the Spark UI
+    spark = SparkSession.builder.appName('clean').getOrCreate()
+
+    # Whether to backfill data to GCS bucket
+    upload = True
```

Comment:
It might be better to get the upload value from the arguments. For example, if somebody adds a --test flag, we can make upload = false. We used a similar approach like this in setup.py